documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,183 @@
+
+### Current
+
+```c++
+enum class DiagnosticMessageType : uint32_t
+{
+    // EventPipe
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+
+### Proposed
+
+Connection:
+```
+runtime <- client : handshake (set length byte array that represents version)
+runtime -> client : sends handshake back as ACK
+runtime <- client : packet containing start code
+runtime -> client : ACK + potential payload
+repeat till closed
+```
+
+### Transport considerations
+
+* All data will be sent in the endianess of the machine
+* All data will be sent in a packet containing a header of information concerning the contents
+
+### Handshake
+
+#### Requirements:
+* simple - easy to code
+* fast - no complex structures/serialization
+* extensible - leaves room to change resulting behavior
+
+The client will initiate the connection by sending a byte array containing the ASCII codes for the version string of the protocol it wishes to communicate with.
+
+> There will only be one for now, but it will give us the ability to extend this and potentially have _other_ protocols we can speak if we'd like.
+
+```c
+// "EVENTPIPE_V1"
+char[15] handshake = { 0x00, 0x00, 0x00, 0x45, 0x56, 0x45, 0x4e, 0x54, 0x50, 0x49, 0x50, 0x45, 0x5f, 0x56, 0x31 };
+```
+
+The `Accept` function in the runtime will immediately try to read 15 bytes of data and then use that determine what protocol we are speaking.  Once it has been determined, the runtime will respond with that handshake.
+
+#### Errors
+* runtime: 
+  * read failure -> end connection; no response
+  * unknown protocol -> ERROR response // TODO: flesh out errors
+* Client:
+  * no response -> simply have a timeout
+  * unkown protocol error -> fix your client
+
+### Headers
+
+#### Requirements:
+* small - shouldn't affect write times or read times
+* simple - flat and blittable
+* descriptive - self describing
+* extensible - ability for us to add funcitonality without necessarily incrementing protocol version
+
+Every packet will start with a header and every header will start with a size.  This makes it easy to read the data directly out of the stream no matter the size or contents of the packet.
+
+> Taking inspiration from Mono's Soft-Debugger protocol, I am splitting command codes into command sets and commands, allowing command codes to be reused between command sets, e.g., 0x01,0x01 and 0x02,0x01 would be different commands despite sharing the same explicit value for the command code.
+
+```c
+// size = 4 + 1 + 1 + 1 = 7 bytes
+struct header
+{
+    uint16 size; // size of packet = size of header + payload
```

Comment:
The current EventPipe start tracing command has a response where the size isn't known up-front (it is the netperf binary format). You may want to model this as a fixed-size response followed by an optional command-specific binary blob. The description of how to parse that blob, including determining its length would be command-specific.

Reply:
This header corresponds to the IPC protocol and not the event pipe packets.  The current event pipe packets would be the opaque payloads of these packets, e.g.,
```
[ [ IPC Header ] [ IPC payload                           ] ]
[ [ IPC Header ] [ [ netperf header] [ netperf payload ] ] ]
```

The read procedure for event pipe would be:
* read first 7 bytes 
* use current logic to read payload

Reply:
OK got it. Is there a particular new capability (value prop) that you are trying to achieve by packetizing the data that way that our current approach does not achieve?

Reply:
The packets are agnostic to what is being transported inside them, so it enables us to send commands with more structure and metadata.  What I have here is really just a formalization of what is already being done.  I suppose that it is not a requirement that communications must _always_ be packetized in this format, e.g., once you have established a connection and sent the first command, all subsequent information may be sent as some other format (like in the case of EventPipe).  That would simplify things, while giving us a structure for expanding on the functionality of the diagnostics server at a later time.

Reply:
> What I have here is really just a formalization of what is already being done.

Maybe I am misunderstanding? As far as I knew what we are doing now is:
runtime <- client: [MessageHeader] [trace start arguments]
client -> runtime: [session id] [raw_netperf_file]


What I thought you wanted to do was take that response data and break it up into fixed size packets with a little header in front of each one. Maybe you were only referring to command and not to the reply?

Reply:
That was my original idea with the intention of making the commands and responses uniform in structure to make them extensible.

The current structure is closer to:
```
R <- C: [MessageHeader] [trace start]
R -> C: [session id]
R -> C: [netperf event block] // essentially a netperf packet
R -> C: [netperf event block] // essentially a netperf packet
...
repeat until the client sends stop command
```

What I proposed initially (when applied to EventPipe):
```
R <- C: [ magic; size; start_stream; reserved ][ optional payload ]
R -> C: [ magic; size; stream_starting; reserved ][ session id]
R -> C: [ magic; size; command; reserved ][netperf event block] 
R -> C: [ magic; size; command; reserved ][netperf event block]
...
repeat until the client sends stop command
```

After feedback, I'm getting the impression that may be over engineered for a 3.0 timeframe, so I think I'm going to scale back the spec a little to get some breathing room.

I think the following would be a good compromise for 3.0:
```
R <- C: [ magic; size; start_stream; reserved ][ optional payload ]
R -> C: [ magic; size; stream_starting; reserved ][ session id]
R -> C: [netperf event block] // essentially a netperf packet
R -> C: [netperf event block] // essentially a netperf packet
...
repeat until the client sends stop command
```
* formalize the packet structure for diagnostic server command/response only in v1
* Every command is a _new_ connection (this is already the case)
* Allow the reuse of the transport for streaming the `netperf` file after a command/response (this is already the case)
* leave other considerations for post 3.0 timeframe

Thoughts?

Reply:
That sounds pretty good to me : ) You would have to determine what the size in the reply packet includes. In the case of your example it probably includes the header + session_id but not anything after that. In a more generalized sense I could imagine every command follows the pattern:

R <- C: [ magic; size; command_id; reserved ] [fixed_size_command_specific_arguments]
C -> R: [ magic; size; command_id; reserved ] [fixed_size_command_specific_response]
R <- C: [optional command-specific trailing input data stream] \<close connection>
R -> C: [optional command-specific trailing output data stream] \<close connection>

Reply:
That's what I was imagining.  Okay, then I will edit this document to reflect this goal for a 3.0 timeframe, and move anything beyond that to a future category in the doc.